File: pkg/dart2cpp/sample/dart/hello_test_final_enhanced.dart
/// Dart 最终增强语法测试示例
///
/// 包含转换器完全支持的更多Dart语法特性
///

void main() {
  print('🔥 Dart 最终增强语法测试开始');

  // 1. 基本数据类型和操作
  testDataTypes();

  // 2. 控制流增强
  testFinalControlFlow();

  // 3. 函数增强
  testFinalFunctions();

  // 4. 集合增强
  testFinalCollections();

  // 5. 类增强
  testFinalClasses();

  // 6. 泛型
  testFinalGenerics();

  print('✅ Dart 最终增强语法测试完成');
}

/// 测试数据类型和操作
void testDataTypes() {
  print('\n--- 数据类型和操作 ---');

  // 基本类型
  int age = 25;
  double price = 99.99;
  String name = '张三';
  bool isStudent = true;

  // 字符串插值
  String info = '姓名: $name, 年龄: $age';

  // 字符串方法
  int strLength = name.length;

  // 数值方法
  String ageStr = age.toString();

  print('数据类型操作完成');
}

/// 测试最终控制流
void testFinalControlFlow() {
  print('\n--- 最终控制流 ---');

  // switch 语句
  int score = 85;
  switch (score) {
    case 100:
    case 90:
      print('优秀');
      break;
    case 80:
      print('良好');
      break;
    default:
      print('其他分数');
  }

  // for 循环
  print('正序循环:');
  for (int i = 1; i <= 3; i++) {
    print('数字: $i');
  }

  // for-in 循环
  List<String> fruits = ['苹果', '香蕉', '橙子'];
  print('水果列表:');
  for (String fruit in fruits) {
    print('  $fruit');
  }

  // 条件表达式
  int testAge = 20;
  String description = testAge >= 18 ? '成年人' : '未成年人';
  print('年龄描述: $description');
}

/// 测试最终函数特性
void testFinalFunctions() {
  print('\n--- 最终函数特性 ---');

  // 简单函数
  String greetName(String name) {
    return '你好, $name!';
  }

  // 多参数函数
  int addNumbers(int a, int b) {
    return a + b;
  }

  // 调用函数
  String greeting = greetName('李四');
  int sum = addNumbers(1, 2);

  print('问候语: $greeting');
  print('两数之和: $sum');
}

/// 测试最终集合特性
void testFinalCollections() {
  print('\n--- 最终集合特性 ---');

  // List 操作
  List<int> numbers = [1, 2, 3, 4, 5];
  print('列表长度: ${numbers.length}');

  // List 方法
  numbers.add(6);
  print('添加元素后长度: ${numbers.length}');

  // Map 操作
  Map<String, int> scores = {'张三': 95, '李四': 87};
  print('映射大小: ${scores.length}');
  print('张三的分数: ${scores['张三']}');
}

/// 测试最终类特性
void testFinalClasses() {
  print('\n--- 最终类特性 ---');

  // 创建对象
  FinalPerson person = FinalPerson('张三', 25);
  print('创建FinalPerson对象');

  // 调用方法
  person.introduce();

  // 访问属性
  print('姓名: ${person.getName()}');

  // 继承
  FinalStudent student = FinalStudent('王五', 20, '计算机科学');
  print('创建FinalStudent对象');
  student.study();
  student.introduce();
}

/// 测试最终泛型特性
void testFinalGenerics() {
  print('\n--- 最终泛型特性 ---');

  // 泛型类实例化
  FinalContainer<String> stringContainer = FinalContainer<String>('Hello');
  FinalContainer<int> intContainer = FinalContainer<int>(42);

  print('字符串容器值: ${stringContainer.getValue()}');
  print('整数容器值: ${intContainer.getValue()}');
}

// ========================
// 类定义
// ========================

/// 最终人物类
class FinalPerson {
  String _name;
  int _age;

  FinalPerson(this._name, this._age);

  // Getter
  String getName() => _name;
  int getAge() => _age;

  /// 自我介绍
  void introduce() {
    print('大家好，我是 $_name，今年 $_age 岁。');
  }

  @override
  String toString() {
    return 'FinalPerson{name: $_name, age: $_age}';
  }
}

/// 最终学生类（继承）
class FinalStudent extends FinalPerson {
  String _major;

  FinalStudent(String name, int age, this._major) : super(name, age);

  // Getter
  String getMajor() => _major;

  /// 学习方法
  void study() {
    print('${getName()} 正在学习 $_major');
  }

  @override
  void introduce() {
    print('大家好，我是 ${getName()}，今年 ${getAge()} 岁，专业是 $_major。');
  }

  @override
  String toString() {
    return 'FinalStudent{name: ${getName()}, age: ${getAge()}, major: $_major}';
  }
}

/// 最终容器类（泛型）
class FinalContainer<T> {
  T _value;

  FinalContainer(this._value);

  T getValue() => _value;
  void setValue(T newValue) => _value = newValue;
}

File: pkg/dart2cpp/sample/cpp_generated/hello_test_final_enhanced.cpp
#include "dart2cpp.h"

// 工具宏定义

// ============================================================================
// 类: FinalPerson
// ============================================================================

class FinalPerson {
private:
  String _name;
  Int _age;
public:
  FinalPerson(String _name, Int _age) : _name(_name), _age(_age) {
  }
  
  String getName() {
    return this->_name;
  }
  
  Int getAge() {
    return this->_age;
  }
  
  Nullable introduce() {
    dart_print(dart_concat(dart_string("大家好，我是 "), (this->_name).toString(), dart_string("，今年 "), (this->_age).toString(), dart_string(" 岁。")));
return Void;
  }
  
  String toString() {
    return dart_concat(dart_string("FinalPerson{name: "), (this->_name).toString(), dart_string(", age: "), (this->_age).toString(), dart_string("}"));
  }
  
};

// ============================================================================
// 类: FinalStudent
// ============================================================================

class FinalStudent : public FinalPerson {
private:
  String _major;
public:
  FinalStudent(String name, Int age, String _major) : _major(_major), FinalPerson(name, age) {
  }
  
  String getMajor() {
    return this->_major;
  }
  
  Nullable study() {
    dart_print(dart_concat((this->getName()).toString(), dart_string(" 正在学习 "), (this->_major).toString()));
return Void;
  }
  
  Nullable introduce() {
    dart_print(dart_concat(dart_string("大家好，我是 "), (this->getName()).toString(), dart_string("，今年 "), (this->getAge()).toString(), dart_string(" 岁，专业是 "), (this->_major).toString(), dart_string("。")));
return Void;
  }
  
  String toString() {
    return dart_concat(dart_string("FinalStudent{name: "), (this->getName()).toString(), dart_string(", age: "), (this->getAge()).toString(), dart_string(", major: "), (this->_major).toString(), dart_string("}"));
  }
  
};

// ============================================================================
// 类: FinalContainer
// ============================================================================

template<typename T>
class FinalContainer {
private:
  T _value;
public:
  FinalContainer(T _value) : _value(_value) {
  }
  
  T getValue() {
    return this->_value;
  }
  
  T setValue(T newValue) {
    return this->_value = newValue;
  }
  
};

Nullable testDataTypes();
Nullable testFinalControlFlow();
Nullable testFinalFunctions();
Nullable testFinalCollections();
Nullable testFinalClasses();
Nullable testFinalGenerics();
Nullable testDataTypes() {
  dart_print(dart_string("\n--- 数据类型和操作 ---"));
auto age = dart_int(25);
auto price = dart_double(99.99);
auto name = dart_string("张三");
auto isStudent = dart_bool(true);
auto info = dart_concat(dart_string("姓名: "), (name).toString(), dart_string(", 年龄: "), (age).toString());
auto strLength = name->size();
auto ageStr = age->toString();
dart_print(dart_string("数据类型操作完成"));
return Void;
}

Nullable testFinalControlFlow() {
  dart_print(dart_string("\n--- 最终控制流 ---"));
auto score = dart_int(85);
if (score == dart_int(100) || score == dart_int(90)) {
  dart_print(dart_string("优秀"));
} else if (score == dart_int(80)) {
  dart_print(dart_string("良好"));
}else {
  dart_print(dart_string("其他分数"));
}
dart_print(dart_string("正序循环:"));
for (auto i = dart_int(1); i->operator_less_equals(dart_int(3)); i = i->operator_add(dart_int(1))) {
dart_print(dart_string("数字: ") + (i).toString());
}
auto fruits = dart_literal<String>(dart_string("苹果"), dart_string("香蕉"), dart_string("橙子"));
dart_print(dart_string("水果列表:"));
auto sync_for_iterator = fruits->iterator();
for (; sync_for_iterator->hasNext(); ) {
auto fruit = sync_for_iterator->next();
dart_print(dart_string("  ") + (fruit).toString());
}
auto testAge = dart_int(20);
auto description = testAge->operator_greater_equals(dart_int(18)) ? dart_string("成年人") : dart_string("未成年人");
dart_print(dart_string("年龄描述: ") + (description).toString());
return Void;
}

Nullable testFinalFunctions() {
  dart_print(dart_string("\n--- 最终函数特性 ---"));
auto greetName = [&](String name) { return dart_concat(dart_string("你好, "), (name).toString(), dart_string("!")); };
auto addNumbers = [&](Int a, Int b) { return a->operator_add(b); };
auto greeting = greetName(dart_string("李四"));
auto sum = addNumbers(dart_int(1), dart_int(2));
dart_print(dart_string("问候语: ") + (greeting).toString());
dart_print(dart_string("两数之和: ") + (sum).toString());
return Void;
}

Nullable testFinalCollections() {
  dart_print(dart_string("\n--- 最终集合特性 ---"));
auto numbers = dart_literal<Int>(dart_int(1), dart_int(2), dart_int(3), dart_int(4), dart_int(5));
dart_print(dart_string("列表长度: ") + (numbers->size()).toString());
numbers->add(dart_int(6));
dart_print(dart_string("添加元素后长度: ") + (numbers->size()).toString());
auto scores = Map<String, Int>::createFromEntries({{dart_string("张三"), dart_int(95)}, {dart_string("李四"), dart_int(87)}});
dart_print(dart_string("映射大小: ") + (scores->size()).toString());
dart_print(dart_string("张三的分数: ") + (scores->operator_index(dart_string("张三"))).toString());
return Void;
}

Nullable testFinalClasses() {
  dart_print(dart_string("\n--- 最终类特性 ---"));
auto person = ObjectPtr<FinalPerson>(new FinalPerson(dart_string("张三"), dart_int(25)));
dart_print(dart_string("创建FinalPerson对象"));
person->introduce();
dart_print(dart_string("姓名: ") + (person->getName()).toString());
auto student = ObjectPtr<FinalStudent>(new FinalStudent(dart_string("王五"), dart_int(20), dart_string("计算机科学")));
dart_print(dart_string("创建FinalStudent对象"));
student->study();
student->introduce();
return Void;
}

Nullable testFinalGenerics() {
  dart_print(dart_string("\n--- 最终泛型特性 ---"));
auto stringContainer = ObjectPtr<FinalContainer<String>>(new FinalContainer<String>(dart_string("Hello")));
auto intContainer = ObjectPtr<FinalContainer<Int>>(new FinalContainer<Int>(dart_int(42)));
dart_print(dart_string("字符串容器值: ") + (stringContainer->getValue()).toString());
dart_print(dart_string("整数容器值: ") + (intContainer->getValue()).toString());
return Void;
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    dart_print(dart_string("🔥 Dart 最终增强语法测试开始"));
testDataTypes();
testFinalControlFlow();
testFinalFunctions();
testFinalCollections();
testFinalClasses();
testFinalGenerics();
dart_print(dart_string("✅ Dart 最终增强语法测试完成"));
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}
